base/shared/src/main/scala/scalaz/control/Inf.scala

```diff
@@ -0,0 +1,43 @@
+package scalaz
+package control
+
+/**
+  * Modelled from http://www.cse.chalmers.se/~nad/publications/danielsson-altenkirch-mixing.pdf
+  */
+sealed abstract class Bool
+sealed abstract class True extends Bool
+sealed abstract class False extends Bool
+
+object Bool {
+  sealed abstract class Value[B <: Bool]
```

Comment:
Should subtyping get in the way (or just get annoying to always mention the bound `<: Bool`), I think everything should work even without this bound (and without the superclass `Bool`).